### PR TITLE
Support external extensions for Crosswalk Webview Android

### PR DESCRIPTION
**Pull request by leonhsl**
Migrate external extensions creation/management framework
from xwalk_app_runtime layer down to xwalk_core_library layer.
Modified extensions-config.json format.
Modified all sample extensions to accommodate with new extension
framework.

BUG=XWALK-5746

### CODE REVIEW

**Comment by leonhsl** [comment]
Design doc: https://docs.google.com/a/intel.com/presentation/d/1sTkf_ITeS_n7NMLfF6iRcGy6hILZWSWP52A_4Yu0i-o/edit?usp=sharing

Manual test all PASSED for App-Runtime/EmbeddingAPI & Embedded/Shared mode.
@halton @huningxin @lincsoon PTAL, Thanks~

**Comment by leonhsl** [comment]
Crosswalk Linux trybot failed but I found that it should be because of unstable trybot. So would you please help start to review? Thanks~ 
@halton @huningxin @lincsoon

**Comment by leonhsl** [comment]
Add @DonnaWuDongxia , Thanks~

**Comment by darktears** [comment]
@pozdnyakov

**Comment by leonhsl** [comment]
Soft ping, Thanks~ @halton @huningxin @lincsoon @pozdnyakov

**Comment by leonhsl** [comment]
Code updated for one thing: Added startActivityForResult() API for XWalkView.

**Comment by halton** [comment]
@leonhsl since #3420 has been merged, this PR need to rework. @DonnaWuDongxia please help to review this PR.

**Comment by darktears** [comment]
And it's also lacking testing.

**Comment by leonhsl** [comment]
Pushed this PR again, PTAL, Thanks~

Updates:
1. Did rework for #3420 
2. Test case PASS: XWalkCoreTest, XWalkRuntimeClientEmbeddedTest, XWalkRuntimeClientTest

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkRuntimeExtensionLoader.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            extList = assetManager.list(XWALK_EXTENSIONS_FOLDER);
+        } catch (IOException e) {
+            Log.w(TAG, "Failed to iterate assets/xwalk-extensions folder");
+            return;
+        }
+
+        for (String path : extList) {
+            // Load the extension.
+            Log.i(TAG, "Start to load extension: " + path);
+            mXWalkView.getExtensionManager().loadExtension(XWALK_EXTENSIONS_FOLDER + "/" + path);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkRuntimeExtensionLoader.java]: Please use `File.seperator`

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkRuntimeExtensionLoader.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    public void loadExtensions() {
+        loadExternalExtensions();
+    }
+
+    private void loadExternalExtensions() {
+        AssetManager assetManager = mActivity.getAssets();
+        String[] extList;
+        try {
+            Log.i(TAG, "Iterate assets/xwalk-extensions folder");
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkRuntimeExtensionLoader.java]: nit, use XWALK_EXTENSIONS_FOLDER to concat the path.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkRuntimeExtensionLoader.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            extList = assetManager.list(XWALK_EXTENSIONS_FOLDER);
+        } catch (IOException e) {
+            Log.w(TAG, "Failed to iterate assets/xwalk-extensions folder");
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkRuntimeExtensionLoader.java]: nit

File: runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java [hunk 4]
```diff
@@ -80,130 +92,142 @@ public Activity getActivity() {
...
+    //     ExtensionA.js(Optional)
     @Override
-    public void postMessage(XWalkExtensionClient extension, int instanceID, String message) {
-        XWalkRuntimeExtensionBridge bridge = mExtensions.get(extension.getExtensionName());
+    public void loadExtension(String extensionPath) {
+        if (!mLoadExternalExtensions) return;
+
+        // Read extension json file.
+        int len = extensionPath.length();
+        if (extensionPath.charAt(len-1) == '/') {
```
Comment (halton) [inline runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java]: space between `-` and use File.seperator

File: runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java [hunk 5]
```diff
@@ -80,130 +92,142 @@ public Activity getActivity() {
...
+            extensionPath = extensionPath.substring(0, len-1);
+        }
+        String folderName = extensionPath.substring(extensionPath.lastIndexOf('/') + 1);
+        String jsonFile = extensionPath + "/" + folderName + ".json";
```
Comment (halton) [inline runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java]: ditto, File.seperator

File: runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java [hunk 6]
```diff
@@ -80,130 +92,142 @@ public Activity getActivity() {
...
+
+        try {
+            JSONObject jsonObject = new JSONObject(jsonFileContent);
+            String name = jsonObject.getString("name");
+            String className =  jsonObject.getString("class");
+            String jsApiFile = jsonObject.optString("jsapi");
+
+            // Load the content of the JavaScript file.
+            if (jsApiFile != null && jsApiFile.length() != 0) {
+                jsApiFile = extensionPath + "/" + jsApiFile;
```
Comment (halton) [inline runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java]: ditto

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkExternalExtensionManagerInternal.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// // Use of this source code is governed by a BSD-style license that can be
+// // found in the LICENSE file.
```
Comment (halton) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkExternalExtensionManagerInternal.java]: One `//` is enough

**Comment by leonhsl** [comment]
@halton Thank you very much for review comments~ Addressed all and pushed again, PTAL, Thanks~

File: runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java [hunk 8]
```diff
@@ -80,130 +92,142 @@ public Activity getActivity() {
...
     @Override
-    public void postMessage(XWalkExtensionClient extension, int instanceID, String message) {
-        XWalkRuntimeExtensionBridge bridge = mExtensions.get(extension.getExtensionName());
+    public void loadExtension(String extensionPath) {
+        if (!mLoadExternalExtensions) return;
+
+        // Read extension json file.
+        int len = extensionPath.length();
+        if (extensionPath.charAt(len - 1) == File.separatorChar) {
+            extensionPath = extensionPath.substring(0, len-1);
```
Comment (halton) [inline runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java]: nit, space around `-`

File: runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java [hunk 9]
```diff
@@ -260,6 +284,23 @@ private void createExternalExtension(String name, String className, String jsApi
...
+        String path = null;
+        try {
+            ApplicationInfo appInfo =
+                    mContext.getPackageManager()
+                    .getApplicationInfo(mActivity.getPackageName(), 0);
+            path = appInfo.nativeLibraryDir;
+        } catch (final NameNotFoundException e) {
+        }
+        if (path != null && new File(path).isDirectory())
+            mNativeExtensionLoader.registerNativeExtensionsInPath(path);
```
Comment (halton) [inline runtime/android/core/src/org/xwalk/core/extension/XWalkExternalExtensionManagerImpl.java]: nit add `{}` for about condition.

**Comment by halton** [comment]
LGTM with the nits. @DonnaWuDongxia @huningxin @pozdnyakov please show your concerns/comments if have.

**Comment by leonhsl** [comment]
@halton  Thank you very much for acknowledgement! Addressed nit issues and pushed again. Thanks~

**Comment by halton** [comment]
LGTM. ping @DonnaWuDongxia @huningxin @pozdnyakov, if no future comments by today, I'll merge it.

**Comment by halton** [comment]
Merging...

**Pull request merged**